xsnap/sources/xsnapPlatform.c

```diff
@@ -300,20 +303,27 @@ void fxQueuePromiseJobs(txMachine* the)
 	the->promiseJobs = 1;
 }
 
+void fxRunDebugger(txMachine* the)
+{
+#ifdef mxDebug
+	fxDebugCommand(the);
+#endif
+}
+
 void fxRunLoop(txMachine* the)
 {
 	c_timeval tv;
 	txNumber when;
 	txJob* job;
 	txJob** address;
-	fxEndJob(the);
```

Comment:
This looks like a potential behavioral change, but I can't remember why we needed this here. I think you added it in 84aa01ac86596d46f98eeb5930bcdd33c66e0ef6 , but the comment just says "Platform runloop tweaks, Snapshot dump tweaks".

I think the command we send down to the worker gets its first turn of execution before we call `fxRunLoop()`. I think `fxEndJob()` basically does finalization callbacks and `fxCheckUnhandledRejections`. We probably don't do a lot during that first turn (https://github.com/Agoric/agoric-sdk/blob/2529e82427277d98ea777c5a35e6bd753e65e6f7/packages/SwingSet/src/supervisors/supervisor-helper.js#L32 does a deliberate stall, which would consume the first turn), but it's conceivable that this might trigger organic GC, and then some finalizers could become ready to run.

I'm ok with this change as long as you can't think of any reason why we need to maintain the previous behavior.

Reply:
The previous behavior was "not expected" and buggy.

If the command job enqueued promises, it would not drain those, but instead run `endJob` first. Only at that point it would drain promise queues (for both jobs)

The expectation is that reactions form an I/O job are processed immediately, then come the "gc" jobs (and its reactions).

The previous addition was necessary because if the main command job didn't queue promises, `endJob` would not trigger. It was actually the wrong fix because the same thing would happen if a subsequent I/O job (e.g. `setImmediate`) didn't queue promises, it wouldn't run `endJob` either.

Now the logic does the expected thing:
- Run the I/O job (command or timers), and drain promise reactions
- Run GC jobs and check for unhandled errors, and drain promise reactions
- Repeat